add a stall detector that logs stacktraces of unyielding tasks, redux

davidblewett opened the pull request:
### What does this PR do?

This PR supercedes https://github.com/DataDog/glommio/pull/417 . I believe all the concerns raised there are addressed in this implementation.

This PR adds a rudimentary stall detection mechanism to Glommio.

It measures the runtime of task queues and logs if they don't yield when
they are due.

Knowing if a queue is stalling the reactor is nice, but finding the
exact code location where the stall occurs remains very hard. Especially
if a task queue hosts many concurrent fibers.

To help with that, this PR introduces a stall detection mechanism that
records stack traces of stalling tasks. It works as follows:

- When a task queue is scheduled for execution, we set up a timer that
triggers some time after the queue is expected to yield (we add a 10%
error margin to avoid false positives). A background thread collocated
with the local executor waits on the timer at all times.
- When the timer fires, the thread sends a signal (SIGUSR1) to the
local executor thread. Upon receiving the signal, the local executor
records a complete trace of the local stack. Here we take advantage of
the fact that by default, the kernel invokes signal handlers on top of
the existing stack. i.e. the frames we record are those of the
problematic user code that was meant to yield. The recorded frames are
pushed on a non-blocking communication channel that links the signal
handler and the local executor.
- When a task queue yields, the local executor disarm the timer and
checks the communication channel for potential recorded frames, if there
are any then we can conclude that the queue stalled, so we log them.

This code works in practice but has two major drawbacks:

- The timer dance is expensive; expect a high number of syscalls.
Because of this runtime overhead, the stall detector is disabled by
default. To opt-in, you  must pass in something that implements
`StallDetectionHandler` when constructing the `LocalExecutor`. There
is a concrete `DefaultStallDetectionHandler` that uses `SIGUSR1` and 
logs a warning log message.
- We log stalls only after the queue yield. Therefore, if there is a bug
in your code and your queue never yields, the stall detector will never
log the code location that's at fault (even though we probably have
recorded the trace by then). The reason for this is that logging from a
signal handler is illegal.

glommer commented on glommio/src/executor/stall.rs:
```diff
@@ -0,0 +1,305 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the MIT/Apache-2.0 License, at your convenience
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2021 Datadog, Inc.
```
2022

glommer commented on glommio/src/executor/stall.rs:
```diff
@@ -0,0 +1,305 @@
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2021 Datadog, Inc.
+//
+#![allow(dead_code)]
```
this is not ideal. If you absolutely must mask something, best to mask it at the use-site level, not the file level

glommer commented on glommio/src/executor/stall.rs:
```diff
@@ -0,0 +1,305 @@
+        start: Instant,
+        max_expected_runtime: Duration,
+    ) -> nix::Result<StallDetectorGuard<'_>> {
+        // We consider a queue to be stalling the system if it failed to yield in due
```
on the other hand for a 500us 10% does little if anything.

The application will know better what threshold to use, so I think it is best to have something user-defined than this heuristic.

HippoBaro replied:
Agreed. @davidblewett let's use the handler object for that like we for deciding which signal to use.

glommer commented on glommio/src/executor/stall.rs:
```diff
@@ -0,0 +1,305 @@
+        threshold: Duration,
+    ) -> nix::Result<Self> {
+        detector.disarm().unwrap();
+        detector.arm(threshold).unwrap();
```
Don't quote me on that because my memory is failing on this, but I am pretty sure we should be able to do a rearm with a single syscall?

disarm + arm, if it does two syscalls, that's a bit much

HippoBaro replied:
Rearming will implicitly override the old timer. IIRC that's how my original work did it and it worked fine

glommer commented:
Generally good, minor comments only. The most important one is the heuristics about when to trigger. I think it is impossible that we'll get this right at the framework level, and the user has to tell us when to fire as they turn this on.

(also CI is not passing)

HippoBaro commented on glommio/src/executor/stall.rs:
```diff
@@ -0,0 +1,305 @@
+
+impl fmt::Debug for StallDetection<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "[stall-detector -- executor {}] task queue {} went over-budget: {:#?} (budget: \
+             {:#?}). Backtrace: {:#?}",
+            self.executor, self.task_queue, self.overage, self.budget, self.trace,
+        )
```
This works but that's not the best way to do this. Here's a better way:
```
fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("RingIoStats")
            .field("executor", &self.executor)
            // ...
            .finish()
}
```

HippoBaro replied:
You should also implement `Display` for this one. And the above is probably a good output for that!

HippoBaro commented on glommio/src/executor/stall.rs:
```diff
@@ -0,0 +1,305 @@
+
+pub struct StallDetection<'a> {
+    executor: usize,
+    task_queue: &'a str,
```
the task queue name is fine, but you definitely want the task queue handle in there as well.

HippoBaro commented on glommio/src/executor/stall.rs:
```diff
@@ -0,0 +1,305 @@
+    timer_handler: Option<JoinHandle<()>>,
+    id: usize,
+    terminated: Arc<AtomicBool>,
+    //NOTE: we don't use signal_hook::low_level::channel as backtraces
+    //      have too many elements
```
nit formatting

HippoBaro commented on glommio/src/executor/stall.rs:
```diff
@@ -0,0 +1,335 @@
+
+    /// The default signal is SIGUSR1.
+    fn signal(&self) -> u8 {
+        nix::libc::SIGUSR1 as u8
```
Same here, let's make this configurable

HippoBaro reviewed (changes requested):
Some nits.

My one concern is that now that I am reading the code, I realize why I used TLS for the channels in my original PR. The reason is that signal handlers are global entities. Therefore I see two issues with this version of the code:

* First, you install a signal handler every time an executor is created. So either (1) it is replaced each time or (2) the library does something clever underneath. Let's make sure it's the former. 
* Second, only a single executor will ever be signaled (the last one), and it will forward the traces to itself (because the handler captures the channels to send to). The only way for this to work is for the signal handler to fetch the channel to send to using TLS. Fortunately, the executor is hosted using TLS, so this should be trivial (Look at `LOCAL_EX`).

I think the best way to install the signal handler is to use some global entirety with a lazy initialization (you need it to be lazy since we can configure what signal to use at run time.)

One last thing... What happens if two executors specify different signals? That's a problem. I say panic in that case. If they are doing that they were looking for trouble.

glommer commented on clippy.toml:
```diff
@@ -0,0 +1 @@
+too-many-arguments-threshold = 8
```
on your comment above, Hippo, maybe the executors shouldn't be able to specify the signal.
If we really want this to be configurable, an env variable is likely the way to go.

glommer replied:
Will this be addressed?

HippoBaro replied:
Yeah this has to go. My work is merged so there is now a proper struct that avoids going over the argument limit.

glommer commented on glommio/Cargo.toml:
```diff
@@ -17,6 +17,7 @@ readme = "../README.md"
 
 [dependencies]
 ahash = "~0.7"
+backtrace = { version = "0.3.61" }
```
in this case we likely need a minimum patchlevel, though. There is a fix that I added to the backtrace crate without which this won't work (although I wouldn't be able to tell you which version has that)

davidblewett replied:
The version turned out to be `0.3.58`. I updated the pin to that minimum.

davidblewett commented on glommio/src/executor/stall.rs:
```diff
@@ -0,0 +1,335 @@
+        // triggers if it doesn't yield after running for 15ms.
+        Some(Duration::from_millis((max_expected_runtime.as_millis() as f64 * 0.1) as u64)
+            .max(Duration::from_millis(10)))
+    }
```
This is the impl for `DefaultStallDetectionHandler`?  I thought our customization method was going to be for someone to implement the `StallDetectionHandler` trait. I'm not a fan of knobs on knobs when it comes to configuration. It becomes increasingly difficult to figure out what exactly is going on. If there's a  `DefaultStallDetectionHandler`, we know exactly what values it's running with.

That being said, I'll bow to consensus if you and @glommer want to make it configurable another layer.

HippoBaro replied:
I don't think calling this the "default" handler is a good idea. Stall detection will be disabled by default and users will have to provide this handler explicitly to enable it. 

I would call it a logging handler, personally since that's what it does. 

As for configuration maybe @glommer has a strong opinion on this. I feel like different applications will have different toleration levels for stalls and so it's not going to be possible for us to give good constants that will work for everyone. 

Instead, I would make it completely customizable and implement `Default`.

davidblewett replied:
Sounds good on renaming and adding the customization bits.

davidblewett commented:
> My one concern is that now that I am reading the code, I realize why I used TLS for the channels in my original PR. The reason is that signal handlers are global entities. Therefore I see two issues with this version of the code:
> 
> * First, you install a signal handler every time an executor is created. So either (1) it is replaced each time or (2) the library does something clever underneath. Let's make sure it's the former.

The library does (2). From the [docs](https://docs.rs/signal-hook/latest/signal_hook/low_level/fn.register.html): ```This makes sure there’s a signal handler for the given signal. It then adds the action to the ones called each time the signal is delivered. If multiple actions are set for the same signal, all are called, in the order of registration.```

> * Second, only a single executor will ever be signaled (the last one), and it will forward the traces to itself (because the handler captures the channels to send to). The only way for this to work is for the signal handler to fetch the channel to send to using TLS. Fortunately, the executor is hosted using TLS, so this should be trivial (Look at `LOCAL_EX`).

I added a unit test that exercises this behavior. The appropriate thread receives the signal. I also added checks to make sure that the signal handler runs in the expected thread, to handle the case where the process as a whole receives the signal and the OS picked our thread to handle it.

> I think the best way to install the signal handler is to use some global entirety with a lazy initialization (you need it to be lazy since we can configure what signal to use at run time.)

I'd like to stick with the current approach; I think the abstractions provided by `signal_hook` provide the functionality we need.

> One last thing... What happens if two executors specify different signals? That's a problem. I say panic in that case. If they are doing that they were looking for trouble.

I added a unit test exercising this. I don't think it's a problem with the implementation as is.

HippoBaro commented on glommio/src/executor/stall.rs:
```diff
@@ -0,0 +1,543 @@
+    /// What signal number to use; see values in libc::SIG*
+    fn signal(&self) -> u8;
+
+    /// Handler called when a task exceeds its budget
+    fn stall(&self, detection: StallDetection<'_>);
```
Let's add a comment to explain this callback is invoked from inside the reactor. We should ask users to not perform blocking syscalls, not to acquire locks, not to spawn tasks, etc.

HippoBaro replied:
In fact, I would be almost tempted to unset the LOCAL_EX TLS var just so that trying to enter the executor would panic the app.

HippoBaro commented on glommio/src/executor/stall.rs:
```diff
@@ -0,0 +1,543 @@
+                        .is_ok()
+                });
+            })?;
+            //.map_err::<io::Error, GlommioError<()>>(GlommioError::from)?;
```
nit

HippoBaro replied:
Btw I think you can just return the directly rather than using `?` + `Ok(signal_id)`

HippoBaro commented on glommio/src/executor/stall.rs:
```diff
@@ -0,0 +1,543 @@
+        time::Duration,
+    };
+
+    #[allow(dead_code)]
```
I don't think this is necessary?

davidblewett replied:
Not all of the fields are used by tests currently. I can remove the fields if it's a problem?

HippoBaro replied:
If you plan to add tests that use this then let's wait on those; otherwise, let's remove it :)

HippoBaro commented on glommio/src/executor/stall.rs:
```diff
@@ -0,0 +1,543 @@
+                // no stall because < 50ms of un-cooperativeness
+                thread::sleep(Duration::from_millis(40));
+
+                assert!(stall_handler.inner.read().unwrap().detections.is_empty());
```
This will never trigger. Not because you have not stalled but because the handler is called after the task queue has returned. Since you are executing in the queue, it will always be empty.

You need to move this assertion after the yielding point just below.

HippoBaro commented on glommio/src/executor/stall.rs:
```diff
@@ -0,0 +1,543 @@
+            build_handlers.push((handler, builder));
+        }
+        let mut handles = Vec::with_capacity(10);
+        for (handler, builder) in build_handlers.drain(..) {
```
Not an issue but FYI you don't need to drain here, just iterate on `build_handlers` and that will consume it.

HippoBaro reviewed (changes requested):
Some very final comments but overall this is looking really good

HippoBaro commented on glommio/src/executor/mod.rs:
```diff
@@ -680,6 +715,32 @@ pub struct LocalExecutorPoolBuilder {
+
+impl fmt::Debug for LocalExecutorPoolBuilder {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let gen = self
+            .handler_gen
+            .as_ref()
+            .map(|_| "Some(Box<dyn Fn() -> Box<dyn stall::StallDetectionHandler + 'static>)");
+        write!(
+            f,
+            "spin_before_park:{:?}, name:{}, io_memory:{}, ring_depth:{}, \
+             preempt_timer_duration:{:?}, placement:{:?}, record_io_latencies:{}, handler_gen:{:?}",
+            self.spin_before_park,
+            self.name,
+            self.io_memory,
+            self.ring_depth,
+            self.preempt_timer_duration,
+            self.placement,
+            self.record_io_latencies,
+            gen
```
bump on this; it was marked resolved but it hasn't changed. Maybe there was a mixup :shrug:

glommer commented on glommio/src/executor/stall.rs:
```diff
@@ -0,0 +1,534 @@
+    ) -> Option<Duration>;
+
+    /// What signal number to use; see values in libc::SIG*
+    fn signal(&self) -> u8;
```
why don't those have default implementations ?
I think it is best for a user to usually only specify the water mark, the rest I'd expect to be defaults most of the time.

HippoBaro replied:
I didn't think about that, I agree it would be best to provide an overridable default implementation for that

davidblewett replied:
I can see providing a default for `signal` (`SIGUSR1`), but I don't see what a default would be for the `stall` function? There could be any number of actions someone might want to take. The logging one just logs, but we're probably going to add one that emits Datadog events or something along those lines.

glommer replied:
That just means that the one that emits Datadog events won't use the default =)
Logging is a great default.

glommer commented on glommio/src/executor/stall.rs:
```diff
@@ -0,0 +1,534 @@
+pub trait StallDetectionHandler: std::fmt::Debug + Send + Sync {
+    /// How far past the preemption timer should qualify as a stall
+    /// If None is returned, don't use the stall detector for this task queue.
+    fn high_water_mark(
```
this name feels weird to me, maybe it is an appendage of earlier design?
Usually a high watermark implies a low watermark...

HippoBaro replied:
I take the blame here, I suggested the name and I suck at names :laughing: Maybe `threshold` would work better for you?

glommer replied:
something like `trigger_threshold` or `threshold` would indeed be better, but ultimately I don't have a problem with watermark. It's the `high` part that feels weird without a corresponding low

davidblewett replied:
I'll go with `threshold`.

glommer commented on glommio/src/executor/stall.rs:
```diff
@@ -0,0 +1,534 @@
+    pub signal: u8,
+    /// What percentage of the original expected runtime to allow before
+    /// triggering a stall
+    pub overage_ratio: f64,
```
I am still not a fan of the float ratio.
My reason for that is that it is hard to reason about.

I see where you're coming from: because there can't be different latency goals, you'd likely be okay with a 30ms overshoot on a 100ms latency goal and only okay with a 3ms overshoot on a 10ms goal.

But there are two big problems with that:
1) It is essentially impossible to predict what the goal will be. The latency threshold is not a property of a task queue: it is a property of a *set of runnable task queues*. If there is a (runnable) task queue with a 100ms threshold and another with 1ms, then the threshold is 1ms. This comes and goes and I think it complicates the design.

2) In an ideal world, nobody ever goes over the task quota, at any time, for no period of time. (or an absolute threshold == 0) Of course such world doesn't exist, so we log those events. But it doesn't sound okay to me to overshoot the quota for 30ms just because your task queue has a higher threshold. Ultimately, the stall detector is not about your task queue, but about the fact that you're preventing other task queues to run by clogging the pipes.

glommer commented on glommio/src/executor/stall.rs:
```diff
@@ -0,0 +1,534 @@
+}
+
+impl LoggingStallDetectionHandler {
+    fn new(signal: u8, overage_ratio: f64, min_duration: Duration) -> Self {
```
I think we should provide a concrete class with no settable properties, and allow extensions only through traits.

HippoBaro replied:
I was pushing for a configurable class that provides `Default`

glommer reviewed (commented):
Very minor comments at this point. 
In the trait vs tunables, I prefer the traits and you seem too as well, so let's do that.

Let me know what you think about my argument against the percentage-based threshold.
I think we should provide a concrete class with, say, 10ms flat default and then allow overriding through traits.

davidblewett commented:
> Let me know what you think about my argument against the percentage-based threshold. I think we should provide a concrete class with, say, 10ms flat default and then allow overriding through traits.

This seems reasonable to me. It's very easy to implement the exact behavior you want if this doesn't fit your use case.

davidblewett commented:
The test failures are due to newer lints in rust 1.58+, that seem unrelated to this PR.

HippoBaro commented on glommio/src/executor/mod.rs:
````diff
@@ -566,6 +578,28 @@ impl LocalExecutorBuilder {
         self
     }
 
+    /// Whether to detect stalls in unyielding tasks.
+    /// [`DefaultStallDetectionHandler`] installs a signal handler for
+    /// [`nix::libc::SIGUSR1`], so is disabled by default.
+    /// # Examples
+    ///
+    /// ```
+    /// use glommio::{DefaultStallDetectionHandler, LocalExecutorBuilder};
+    ///
+    /// let local_ex = LocalExecutorBuilder::default()
+    ///     .detect_stalls(Some(Box::new(DefaultStallDetectionHandler {})))
+    ///     .make()
+    ///     .unwrap();
+    /// ```
````
Can you use the same doc as in there? https://github.com/DataDog/glommio/pull/499/files#diff-5b724ff10c950f45f86b8b247e7b6000f5f6cba737c7ca8e2ebe1ec5e85bc560R876-R899

davidblewett replied:
But they're different? `LocalExecutorBuilder` only takes the stall handler directly. `LocalExecutorPoolBuilder` takes a handler factory function, and its doc uses `on_all_shards` which isn't on `LocalExecutorBuilder`?

HippoBaro commented on glommio/src/executor/mod.rs:
````diff
@@ -1060,7 +1155,31 @@ impl LocalExecutor {
+    /// let local_ex =
+    ///     LocalExecutor::default().detect_stalls(Some(Box::new(DefaultStallDetectionHandler {})));
+    /// ```
+    pub fn detect_stalls(
+        &mut self,
+        handler: Option<Box<dyn stall::StallDetectionHandler + 'static>>,
+    ) -> Result<()> {
+        self.stall_detector = match handler {
+            Some(handler) => Some(StallDetector::new(self.id, handler)?),
+            None => None,
+        };
+        Ok(())
````
Can you add one final test that ensures this is safe for `pub` visibility? i.e., that we can safely disable it from outside the reactor? 

I'm asking because while user code is touching this, the stall detector is actively used, and a timer exists. I want to ensure that disabling it doesn't result in anything weird. 

Two little tests should do:
* A test with a cooperative task that disables the stall detector mid-course -> assert that we don't record anything;
* A test with an unyielding task that disables the stall detector mid-course -> assert that we don't produce a trace (or that we do one last time, depending on what behavior is easier to implement).

davidblewett replied:
I removed the export for now.

HippoBaro commented on glommio/src/sys/uring.rs:
```diff
@@ -1040,7 +1040,13 @@ impl SleepableRing {
             // it successfully to be able to safely sleep.
 
-            if self.ring.submit_sqes()? != 1 {
+            if self
+                .submit_sqes()
+                .or_else(Reactor::busy_ok)
+                .or_else(Reactor::again_ok)
+                .or_else(Reactor::intr_ok)?
+                != 1
+            {
```
This is in master so it needs to go, please drop the commit :)

HippoBaro commented on glommio/src/sys/uring.rs:
```diff
@@ -1049,13 +1055,22 @@ impl SleepableRing {
             } else {
                 // The rings are linked. Goodnight!
-                self.ring.cq().wait(1).map(|_| 1)
+                self.ring
+                    .cq()
+                    .wait(1)
+                    .map(|_| 1)
+                    .or_else(Reactor::busy_ok)
+                    .or_else(Reactor::again_ok)
+                    .or_else(Reactor::intr_ok)
```
ditto

HippoBaro commented on glommio/src/sys/uring.rs:
```diff
@@ -1049,13 +1055,22 @@ impl SleepableRing {
             // Submit what we have, once we're out of here we'll consume them
             // and at some point will be able to sleep again.
             self.submit_sqes()
+                .or_else(Reactor::busy_ok)
+                .or_else(Reactor::again_ok)
+                .or_else(Reactor::intr_ok)
```
ditto

HippoBaro commented on glommio/src/sys/uring.rs:
```diff
@@ -1548,9 +1563,9 @@ impl Reactor {
+    fn intr_ok<T: Default>(x: std::io::Error) -> io::Result<T> {
         match x.raw_os_error() {
-            Some(libc::EINTR) => Ok(0),
+            Some(libc::EINTR) => Ok(Default::default()),
```
ditto

HippoBaro reviewed (changes requested):
Almost there!!
